les/vflux/server/service.go

```diff
@@ -0,0 +1,117 @@
+// Copyright 2020 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package server
+
+import (
+	"net"
+	"strings"
+	"time"
+
+	"github.com/ethereum/go-ethereum/les/utils"
+	"github.com/ethereum/go-ethereum/les/vflux"
+	"github.com/ethereum/go-ethereum/p2p/enode"
+	"github.com/ethereum/go-ethereum/rlp"
+)
+
+type (
+	// Server serves vflux requests
+	Server struct {
+		limiter         *utils.Limiter
+		services        map[string]*serviceEntry
+		delayPerRequest time.Duration
+	}
+
+	// Service is a service registered at the Server and identified by a string id
+	Service interface {
+		ServiceInfo() (id, desc string)                                      // only called during registration
+		Handle(id enode.ID, address string, name string, data []byte) []byte // never called concurrently
+	}
+
+	serviceEntry struct {
+		id, desc string
+		backend  Service
+	}
+)
+
+// NewServer creates a new Server
+func NewServer(delayPerRequest time.Duration) *Server {
+	return &Server{
+		limiter:         utils.NewLimiter(1000),
+		delayPerRequest: delayPerRequest,
+		services:        make(map[string]*serviceEntry),
+	}
+}
+
+// Register registers a Service
+func (s *Server) Register(b Service) {
+	srv := &serviceEntry{backend: b}
+	srv.id, srv.desc = b.ServiceInfo()
+	if strings.Contains(srv.id, ":") {
+		// srv.id + ":" will be used as a service database prefix
+		panic("Service ID contains ':'")
```

Comment:
Let's not do the panic.... Return directly is better.

Reply:
done